### DISCUSSION TITLE
variable with scope of entire workflow

### DISCUSSION THREAD

**Discussion by teodik** [community/community#113228] [Actions] [answered]
### Select Topic Area

Question

### Body

I need to create a variable in one job and use it in other jobs. for testing I wrote below code but it doesn't work?

name: Dynamic Variable Example
on: [workflow_dispatch]

jobs:
  set-variable:
    runs-on: ubuntu-latest
    steps:
      - name: Set Dynamic Variable
        id: setvar
        run: |
          echo "MY_DYNAMIC_VARIABLE=value" >> $GITHUB_ENV
          echo "::set-output name=mydynamicvar::value"  # Set an output for use in the next job

  use-variable:
    runs-on: ubuntu-latest
    needs: set-variable
    steps:
      - name: Use Dynamic Variable
        run: |
          echo "Dynamic variable value from environment: ${{ env.MY_DYNAMIC_VARIABLE }}"
          echo "Dynamic variable value from output: ${{ needs.set-variable.outputs.mydynamicvar }}"

set-varible job console output is as follow:

Run echo "MY_DYNAMIC_VARIABLE=value" >> $GITHUB_ENV
Warning: The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

use-variable job console output:

Run echo "Dynamic variable value from environment: "
Dynamic variable value from environment: 
Dynamic variable value from output:

**Comment by lin20260407** [comment]
Your workflow is almost correct but seems to be affected by the deprecation of the set-output command, as noted in the warning message. GitHub Actions has updated the way to share data between jobs, favoring the use of environment files over the set-output command. Here's how you can adjust your workflow to use environment files correctly and ensure that the dynamic variable is accessible across jobs:

name: Dynamic Variable Example
on: [workflow_dispatch]

jobs:
  set-variable:
    runs-on: ubuntu-latest
    outputs:
      mydynamicvar: ${{ steps.setvar.outputs.mydynamicvar }}
    steps:
      - name: Set Dynamic Variable
        id: setvar
        run: |
          echo "value" > mydynamicvar.txt
          echo "MY_DYNAMIC_VARIABLE=value" >> $GITHUB_ENV
          echo "mydynamicvar=value" >> $GITHUB_ENV # Use environment files to share data

  use-variable:
    runs-on: ubuntu-latest
    needs: set-variable
    steps:
      - name: Use Dynamic Variable
        run: |
          echo "Dynamic variable value from environment: ${{ env.MY_DYNAMIC_VARIABLE }}"
          echo "Dynamic variable value from output: ${{ needs.set-variable.outputs.mydynamicvar }}"

In this revised workflow:

Setting the Variable: In the set-variable job, instead of using ::set-output, you now directly write to the $GITHUB_ENV file to set environment variables. These variables become available to all subsequent steps in that job. To pass variables between jobs, you define an output at the job level (outputs.mydynamicvar) and assign the value directly from the environment variable. Note that you must still declare outputs at the job level to make them accessible to other jobs.

Accessing the Variable: In the use-variable job, you can access the dynamic variable in two ways: through the environment variable $MY_DYNAMIC_VARIABLE (which is set but would only be accessible within the same job if you were to use it there) and through the job output needs.set-variable.outputs.mydynamicvar.

Note on Deprecation: The warning message you received is because GitHub Actions is moving away from the set-output command. The revised method uses environment files ($GITHUB_ENV) to set environment variables and outputs at the job level for cross-job variable sharing.

Remember, for multi-line scripts, you might need to use | for YAML multiline strings, and for environment variables to persist between steps, they must be appended to the $GITHUB_ENV file.

I hope I can help you.

**Reply by teodik** [reply]
Copied and pasted the suggested code but variable didn't pass on:

Run echo "Dynamic variable value from environment: "
Dynamic variable value from environment: 
Dynamic variable value from output:

**Reply by lin20260407** [reply]
you mean you wana create a golbal variable and let it can be used in your action file's any place, such as stage1 stage2 all can use ?

**Reply by teodik** [reply]
in my workflow at one of the job I will get an access token to 3rd party tool. In following jobs I need to use that token to access the tool.

**Reply by lin20260407** [reply, marked as answer]
try this my friend 

name: Dynamic Variable Example
on: [workflow_dispatch]

jobs:
  set-variable:
    runs-on: ubuntu-latest
    outputs:
      mydynamicvar: ${{ steps.setvar.outputs.mydynamicvar }}
    steps:
      - name: Set Dynamic Variable
        id: setvar
        run: |
          echo "value" > mydynamicvar.txt
          echo "MY_DYNAMIC_VARIABLE=value" >> $GITHUB_ENV
          echo "::set-output name=mydynamicvar::$(<mydynamicvar.txt)"

  use-variable:
    runs-on: ubuntu-latest
    needs: set-variable
    steps:
      - name: Use Dynamic Variable
        run: |
          echo "Dynamic variable value from output: ${{ needs.set-variable.outputs.mydynamicvar }}"

**Reply by teodik** [reply]
it worked thanks:

Run echo "Dynamic variable value from output: value"
Dynamic variable value from output: value

**Reply by lin20260407** [reply]
Nice job bro
